core/arch/arm/kernel/tee_ta_manager.c

```diff
@@ -236,373 +242,224 @@ static struct tee_ta_ctx *tee_ta_context_find(const TEE_UUID *uuid)
-	if ((sub_head->flags & opt_flags) != sub_head->flags ||
-	    (sub_head->flags & man_flags) != man_flags) {
-		EMSG("TA flag issue: flags=%x opt=%X man=%X",
-		     sub_head->flags, opt_flags, man_flags);
-		return TEE_ERROR_BAD_FORMAT;
+	ctx->mm = tee_mm_alloc(&tee_mm_sec_ddr, vasize);
+	if (!ctx->mm) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
 	}
 
-	ctx->flags = sub_head->flags;
-
-	/* Check if multi instance && single session config  */
-	if (((ctx->flags & TA_FLAG_SINGLE_INSTANCE) == 0) &&
-	    ((ctx->flags & TA_FLAG_MULTI_SESSION) == 0)) {
-		/*
-		 * assume MultiInstance/SingleSession,
-		 * same as MultiInstance/MultiSession
-		 */
-		ctx->flags |= TA_FLAG_MULTI_SESSION;
+	/* Currently all TA must execute from DDR */
+	if (!(ta_head->flags & TA_FLAG_EXEC_DDR)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
 	}
+	ctx->flags = TA_FLAG_EXEC_DDR;
 
 	/* Ensure proper aligment of stack */
-	ctx->stack_size = ROUNDUP(sub_head->stack_size,
-				      TEE_TA_STACK_ALIGNMENT);
-
-	*heap_size = sub_head->heap_size;
+	ctx->stack_size = ROUNDUP(ta_head->stack_size,
+				  TEE_TA_STACK_ALIGNMENT);
 
-	/*
-	 * Allocate heap and stack
-	 */
-	ctx->mm_heap_stack = tee_mm_alloc(&tee_mm_sec_ddr,
-					*heap_size + ctx->stack_size);
-	if (!ctx->mm_heap_stack) {
-		EMSG("Failed to allocate %zu bytes\n",
-			*heap_size + ctx->stack_size);
-		EMSG("  of memory for user heap and stack\n");
-		return TEE_ERROR_OUT_OF_MEMORY;
+	ctx->mm_stack = tee_mm_alloc(&tee_mm_sec_ddr, ctx->stack_size);
+	if (!ctx->mm_stack) {
+		EMSG("Failed to allocate %zu bytes for user stack",
+		     ctx->stack_size);
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
 	}
 
 	/*
-	 * Note that only User TA can be supported in DDR
-	 * if executing in DDR, the size of the execution area
-	 */
-	size += sizeof(ta_head_t) +
-		sec_head->nbr_func * sizeof(ta_func_head_t) +
-		(sec_head->rel_dyn_got_size & TA_HEAD_GOT_MASK);
-
-	ctx->mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
-	if (!ctx->mm)
-		return TEE_ERROR_OUT_OF_MEMORY;
-
-	/*
-	 * Copy TA into reserved memory space (DDR).
+	 * Map physical memory into TA virtual memory
 	 */
 
 	res = tee_mmu_init(ctx);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
 
 	res = tee_mmu_map(ctx, &param);
-	if (res != TEE_SUCCESS) {
-		EMSG("call tee_mmu_map_uta() failed %X", res);
-		return res;
-	}
+	if (res != TEE_SUCCESS)
+		goto out;
 
 	tee_mmu_set_ctx(ctx);
 
-	dst = (void *)tee_mmu_get_load_addr(ctx);
-	if (!tee_vbuf_is_non_sec(nmem_ta, size)) {
-		EMSG("User TA isn't in non-secure memory");
-		return TEE_ERROR_SECURITY;
+	res = elf_load_body(elf_state, tee_mmu_get_load_addr(ctx));
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	digest = malloc(shdr->hash_size);
+	if (!digest) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
 	}
-	memcpy(dst, nmem_ta, size);
 
-	cache_maintenance_l1(DCACHE_AREA_CLEAN, dst, size);
-	cache_maintenance_l1(ICACHE_AREA_INVALIDATE, dst, size);
+	res = crypto_ops.hash.final(hash_ctx, hash_algo, digest,
```

Comment:
Can't we have the `final` and `memcmp` right after the `crypto_ops.hash.update`? In that case, can't we have it part of `tee_ta_load_check_shdr`?